File: automation/localizacao.py
"""
Filtro de candidatura por MODALIDADE (remoto/híbrido/presencial) + REGIÃO do Brasil.

Regra (pedido do usuário):
- O candidato marca no dashboard quais modalidades aceita e, para presencial/híbrido,
  as REGIÕES do Brasil onde toparia trabalhar/mudar (qualquer cidade daquela região).
- Vaga REMOTA → candidata se 'remoto' está aceito (sem restrição de local).
- Vaga PRESENCIAL/HÍBRIDA → candidata só se a modalidade está aceita E a cidade da
  vaga cai numa das regiões aceitas. Ex.: Joinville (SC → Sul) presencial/híbrida →
  candidata se 'Sul' está nas regiões.

Política FAIL-OPEN (igual ao resto do projeto): na dúvida, candidata. Se nada foi
configurado, ou a modalidade/região da vaga não é identificável, NÃO bloqueia.

Módulo de LÓGICA PURA (sem Selenium/DOM) — testável isoladamente.
"""

import re

# UF → região (IBGE).
UF_REGIAO = {
    "AC": "Norte", "AP": "Norte", "AM": "Norte", "PA": "Norte", "RO": "Norte",
    "RR": "Norte", "TO": "Norte",
    "AL": "Nordeste", "BA": "Nordeste", "CE": "Nordeste", "MA": "Nordeste",
    "PB": "Nordeste", "PE": "Nordeste", "PI": "Nordeste", "RN": "Nordeste", "SE": "Nordeste",
    "DF": "Centro-Oeste", "GO": "Centro-Oeste", "MT": "Centro-Oeste", "MS": "Centro-Oeste",
    "ES": "Sudeste", "MG": "Sudeste", "RJ": "Sudeste", "SP": "Sudeste",
    "PR": "Sul", "RS": "Sul", "SC": "Sul",
}

# Nome do estado por extenso → UF (locations às vezes vêm sem a sigla).
ESTADO_UF = {
    "acre": "AC", "amapá": "AP", "amapa": "AP", "amazonas": "AM", "pará": "PA", "para": "PA",
    "rondônia": "RO", "rondonia": "RO", "roraima": "RR", "tocantins": "TO",
    "alagoas": "AL", "bahia": "BA", "ceará": "CE", "ceara": "CE", "maranhão": "MA",
    "maranhao": "MA", "paraíba": "PB", "paraiba": "PB", "pernambuco": "PE", "piauí": "PI",
    "piaui": "PI", "rio grande do norte": "RN", "sergipe": "SE",
    "distrito federal": "DF", "goiás": "GO", "goias": "GO", "mato grosso do sul": "MS",
    "mato grosso": "MT",
    "espírito santo": "ES", "espirito santo": "ES", "minas gerais": "MG",
    "rio de janeiro": "RJ", "são paulo": "SP", "sao paulo": "SP",
    "paraná": "PR", "parana": "PR", "rio grande do sul": "RS", "santa catarina": "SC",
}

# Cidades comuns em vagas → UF (capitais + grandes centros e o exemplo Joinville).
# Best-effort: quando a location traz só o nome da cidade, sem a sigla do estado.
CIDADE_UF = {
    "rio branco": "AC", "macapá": "AP", "macapa": "AP", "manaus": "AM", "belém": "PA",
    "belem": "PA", "porto velho": "RO", "boa vista": "RR", "palmas": "TO",
    "maceió": "AL", "maceio": "AL", "salvador": "BA", "fortaleza": "CE", "são luís": "MA",
    "sao luis": "MA", "joão pessoa": "PB", "joao pessoa": "PB", "recife": "PE",
    "teresina": "PI", "natal": "RN", "aracaju": "SE",
    "brasília": "DF", "brasilia": "DF", "goiânia": "GO", "goiania": "GO",
    "campo grande": "MS", "cuiabá": "MT", "cuiaba": "MT",
    "vitória": "ES", "vitoria": "ES", "belo horizonte": "MG", "uberlândia": "MG",
    "uberlandia": "MG", "rio de janeiro": "RJ", "são paulo": "SP", "sao paulo": "SP",
    "campinas": "SP", "santos": "SP", "são josé dos campos": "SP", "sao jose dos campos": "SP",
    "curitiba": "PR", "londrina": "PR", "maringá": "PR", "maringa": "PR",
    "porto alegre": "RS", "caxias do sul": "RS", "florianópolis": "SC", "florianopolis": "SC",
    "joinville": "SC", "blumenau": "SC", "são josé": "SC", "sao jose": "SC",
}

_MODALIDADES_VALIDAS = ("remoto", "hibrido", "presencial")


def _norm(s: str) -> str:
    return re.sub(r"\s+", " ", (s or "").lower()).strip()


def modalidade_do_texto(texto: str) -> str:
    """Detecta a modalidade da vaga a partir do texto (modalidade/local/descrição).
    Ordem importa: 'híbrido' costuma citar remoto+presencial, então vem primeiro.
    Retorna 'remoto'|'hibrido'|'presencial'|'' (desconhecida)."""
    t = _norm(texto)
    if any(k in t for k in ("híbrid", "hibrid", "hybrid", "semipresencial", "semi-presencial")):
        return "hibrido"
    if any(k in t for k in ("remoto", "remote", "home office", "home-office", "trabalho remoto",
                            "100% remoto", "anywhere", "teletrabalho")):
        return "remoto"
    if any(k in t for k in ("presencial", "on-site", "on site", "onsite", "no local",
                            "no escritório", "no escritorio", "in office", "in-office")):
        return "presencial"
    return ""


def extrair_uf(texto: str) -> str:
    """Extrai a UF (sigla) do texto de localização. Tenta: sigla isolada (', SC' /
    '- SC' / '/SC' / '(SC)'), nome do estado por extenso, e por fim a cidade conhecida.
    Retorna '' se não achar."""
    if not texto:
        return ""
    bruto = texto
    t = _norm(texto)

    # 1) Sigla de UF isolada, cercada por vírgula/hífen/barra/parênteses/espaço/fim.
    for uf in UF_REGIAO:
        if re.search(rf"(?:^|[\s,\-/(]){uf}(?:[\s,\-/)\.]|$)", bruto, flags=re.IGNORECASE):
            return uf

    # 2) Nome do estado por extenso (casa o mais longo primeiro p/ não pegar
    #    'mato grosso' dentro de 'mato grosso do sul').
    for nome in sorted(ESTADO_UF, key=len, reverse=True):
        if nome in t:
            return ESTADO_UF[nome]

    # 3) Cidade conhecida (mesma ordenação por tamanho).
    for cidade in sorted(CIDADE_UF, key=len, reverse=True):
        if cidade in t:
            return CIDADE_UF[cidade]
    return ""


def regiao_do_texto(texto: str) -> str:
    """Região do Brasil da vaga (via UF) OU a própria região citada no texto.
    Retorna '' quando indeterminada."""
    uf = extrair_uf(texto)
    if uf:
        return UF_REGIAO.get(uf, "")
    # A vaga pode citar a região diretamente ('vaga na região Sul').
    t = _norm(texto)
    for reg in ("centro-oeste", "nordeste", "sudeste", "norte", "sul"):
        if reg in t:
            return reg.title() if reg != "centro-oeste" else "Centro-Oeste"
    return ""


def _norm_regioes(regioes) -> set:
    return {_norm(r).replace("centro oeste", "centro-oeste") for r in (regioes or []) if _norm(r)}


def vaga_aceita(texto_vaga: str, modalidades_aceitas, regioes) -> tuple:
    """Decide se deve candidatar-se, dado o texto da vaga e as prefs do candidato.

    Retorna (aceita: bool, motivo: str). FAIL-OPEN: sem config ou sem sinal claro,
    candidata (True)."""
    mods = {_norm(m) for m in (modalidades_aceitas or []) if _norm(m)}
    if not mods:
        return True, "sem filtro de modalidade configurado"

    mod = modalidade_do_texto(texto_vaga)
    if not mod:
        return True, "modalidade da vaga indefinida (fail-open)"
    if mod not in mods:
        return False, f"modalidade '{mod}' não aceita"
    if mod == "remoto":
        return True, "remoto aceito"

    # presencial/híbrido → depende da região da vaga.
    regs = _norm_regioes(regioes)
    if not regs:
        return True, f"{mod} aceito (sem região configurada, fail-open)"
    reg = regiao_do_texto(texto_vaga)
    if not reg:
        return True, f"{mod}: região da vaga indefinida (fail-open)"
    if _norm(reg) in regs:
        return True, f"{mod} na região {reg} (aceita)"
    return False, f"{mod} fora das regiões aceitas (vaga em {reg})"

File: extension/lib/localizacao.js
// Filtro por MODALIDADE (remoto/híbrido/presencial) + REGIÃO — porta de
// automation/localizacao.py. Lógica pura. Exposto em OA.vagaAceita.
// FAIL-OPEN: sem config ou sinal indefinido → candidata (não bloqueia).
(function () {
  const OA = window.OA;
  if (!OA || OA.vagaAceita) return;

  const UF_REGIAO = {
    AC: "norte", AP: "norte", AM: "norte", PA: "norte", RO: "norte", RR: "norte", TO: "norte",
    AL: "nordeste", BA: "nordeste", CE: "nordeste", MA: "nordeste", PB: "nordeste", PE: "nordeste", PI: "nordeste", RN: "nordeste", SE: "nordeste",
    DF: "centro-oeste", GO: "centro-oeste", MT: "centro-oeste", MS: "centro-oeste",
    ES: "sudeste", MG: "sudeste", RJ: "sudeste", SP: "sudeste",
    PR: "sul", RS: "sul", SC: "sul",
  };
  const ESTADO_UF = {
    "acre": "AC", "amapá": "AP", "amapa": "AP", "amazonas": "AM", "pará": "PA", "para": "PA", "rondônia": "RO", "rondonia": "RO", "roraima": "RR", "tocantins": "TO",
    "alagoas": "AL", "bahia": "BA", "ceará": "CE", "ceara": "CE", "maranhão": "MA", "maranhao": "MA", "paraíba": "PB", "paraiba": "PB", "pernambuco": "PE", "piauí": "PI", "piaui": "PI", "rio grande do norte": "RN", "sergipe": "SE",
    "distrito federal": "DF", "goiás": "GO", "goias": "GO", "mato grosso do sul": "MS", "mato grosso": "MT",
    "espírito santo": "ES", "espirito santo": "ES", "minas gerais": "MG", "rio de janeiro": "RJ", "são paulo": "SP", "sao paulo": "SP",
    "paraná": "PR", "parana": "PR", "rio grande do sul": "RS", "santa catarina": "SC",
  };
  const CIDADE_UF = {
    "rio branco": "AC", "manaus": "AM", "belém": "PA", "belem": "PA", "porto velho": "RO", "palmas": "TO",
    "maceió": "AL", "maceio": "AL", "salvador": "BA", "fortaleza": "CE", "são luís": "MA", "sao luis": "MA", "joão pessoa": "PB", "joao pessoa": "PB", "recife": "PE", "teresina": "PI", "natal": "RN", "aracaju": "SE",
    "brasília": "DF", "brasilia": "DF", "goiânia": "GO", "goiania": "GO", "campo grande": "MS", "cuiabá": "MT", "cuiaba": "MT",
    "vitória": "ES", "vitoria": "ES", "belo horizonte": "MG", "uberlândia": "MG", "uberlandia": "MG", "rio de janeiro": "RJ", "são paulo": "SP", "sao paulo": "SP", "campinas": "SP", "santos": "SP",
    "curitiba": "PR", "londrina": "PR", "maringá": "PR", "maringa": "PR", "porto alegre": "RS", "caxias do sul": "RS", "florianópolis": "SC", "florianopolis": "SC", "joinville": "SC", "blumenau": "SC",
  };
  const norm = (s) => (s || "").toLowerCase().replace(/\s+/g, " ").trim();

  function modalidadeDoTexto(texto) {
    const t = norm(texto);
    // híbrido ANTES de presencial ("semipresencial" contém "presencial")
    if (/(híbrid|hibrid|hybrid|semipresencial|semi-presencial|modelo\s+híbrido|modelo\s+hibrido)/.test(t)) return "hibrido";
    if (/(remoto|remote|home[- ]office|anywhere|teletrabalho|trabalho (100%\s*)?remoto|vaga remota|100%\s*remot)/.test(t)) return "remoto";
    if (/(presencial|on[- ]?site|no local|no escritório|no escritorio|in office|trabalho presencial|vaga presencial|100%\s*presencial)/.test(t)) return "presencial";
    return "";
  }

  // Nível do texto (TÍTULO da vaga OU senioridade do perfil) → junior|pleno|senior|gestao.
  // GESTÃO primeiro: "Gerente de Desenvolvimento" contém palavra de dev mas é cargo de
  // CHEFIA — um candidato técnico NÃO deve aplicar. Depois sênior: "Líder/Especialista"
  // não pode ser mascarado por um "júnior" no meio do texto; \b evita casar "jr" em palavra.
  function nivelDoTexto(t) {
    const s = norm(t);
    if (!s) return "";
    if (/(\bgerente\b|\bmanager\b|coordenador|\bhead of\b|\bdiretor|supervisor)/.test(s)) return "gestao";
    if (/(\bs[eê]nior\b|\bsenior\b|\bsr\b\.?|especialista|staff|principal|\blead\b|\bl[íi]der\b|\blider\b|arquitet)/.test(s)) return "senior";
    if (/(\bpleno\b|\bmid\b|mid-level|\bpl\b\.?)/.test(s)) return "pleno";
    if (/(est[aá]gio|trainee|aprendiz|\bj[uú]nior\b|\bjunior\b|\bjr\b\.?)/.test(s)) return "junior";
    return "";
  }
  function extrairUF(texto) {
    if (!texto) return "";
    const t = norm(texto);
    for (const uf of Object.keys(UF_REGIAO)) if (new RegExp(`(?:^|[\\s,\\-/(])${uf}(?:[\\s,\\-/).]|$)`, "i").test(texto)) return uf;
    for (const nome of Object.keys(ESTADO_UF).sort((a, b) => b.length - a.length)) if (t.includes(nome)) return ESTADO_UF[nome];
    for (const cid of Object.keys(CIDADE_UF).sort((a, b) => b.length - a.length)) if (t.includes(cid)) return CIDADE_UF[cid];
    return "";
  }
  function regiaoDoTexto(texto) {
    const uf = extrairUF(texto);
    if (uf) return UF_REGIAO[uf] || "";
    const t = norm(texto);
    for (const r of ["centro-oeste", "nordeste", "sudeste", "norte", "sul"]) if (t.includes(r)) return r;
    return "";
  }

  // (aceita, motivo)
  function vagaAceita(textoVaga, modalidadesAceitas, regioes) {
    const mods = new Set((modalidadesAceitas || []).map(norm).filter(Boolean));
    if (!mods.size) return [true, "sem filtro de modalidade"];
    const mod = modalidadeDoTexto(textoVaga);
    // Modalidade EXPLÍCITA e fora do filtro → descarta na hora (estrito). NÃO
    // identificada → NÃO descarta aqui: muitas páginas não trazem "Remoto" em lugar
    // nenhum e vagas compatíveis estavam sendo jogadas fora ("ignorava vaga que o perfil
    // atende"). Quem fecha esse caso é o MATCH IA, que recebe as modalidades aceitas.
    if (!mod) return [true, "modalidade indefinida — decide a IA"];
    if (!mods.has(mod)) return [false, `modalidade '${mod}' não aceita`];
    if (mod === "remoto") return [true, "remoto aceito"];
    const regs = new Set((regioes || []).map((r) => norm(r).replace("centro oeste", "centro-oeste")).filter(Boolean));
    if (!regs.size) return [true, `${mod} aceito (sem região, fail-open)`];
    const reg = regiaoDoTexto(textoVaga);
    if (!reg) return [true, `${mod}: região indefinida (fail-open)`];
    return regs.has(reg) ? [true, `${mod} na região ${reg}`] : [false, `${mod} fora das regiões (vaga em ${reg})`];
  }

  OA.vagaAceita = vagaAceita;

  // Detecta idioma da vaga (pt|en) por contagem de stopwords — p/ responder no idioma
  // certo (o Selenium fazia via detectar_idioma_texto). Default pt.
  function detectarIdioma(texto) {
    const t = (texto || "").toLowerCase();
    const en = (t.match(/\b(the|and|you|for|with|your|are|will|our|we|is|to|of|in|on|as|experience|requirements|responsibilities|skills|team|work|company|about)\b/g) || []).length;
    const pt = (t.match(/\b(e|de|da|do|para|com|você|voce|sua|são|sao|nossa|nós|nos|é|em|como|experiência|experiencia|requisitos|responsabilidades|habilidades|equipe|trabalho|empresa|vaga|conhecimento|atuar)\b/g) || []).length;
    return en > pt * 1.3 ? "en" : "pt";
  }
  OA.detectarIdioma = detectarIdioma;

  // Gate único usado por TODAS as plataformas:
  //   1) MODALIDADE/REGIÃO (config do usuário) — sinal AMPLIADO: título + descrição +
  //      texto da página. ESTRICTO: modalidade não identificada DESCARTA a vaga
  //      ("se for pleno remoto, apenas vagas pleno e remoto").
  //   2) SENIORIDADE LOCAL (determinística, funciona MESMO se a IA cair) — ESTRICTA:
  //      o nível do título tem que ser IGUAL ao do perfil (pleno↔sênior bloqueia).
  //   3) MATCH IA (limiar por plataforma) — com fallback de modelo (openrouter.js).
  // Retorna {aplicar, motivo}.
  const ESCADA = { junior: 0, pleno: 1, senior: 2, gestao: 3 };
  async function deveAplicar(descricao, { titulo = "", empresa = "", platform = "", pagina = "" } = {}) {
    const cfg = (await OA.bg({ type: "config.get" })).config || {};
    const perfil = cfg.perfil || {};
    const idioma = detectarIdioma([titulo, descricao].filter(Boolean).join("\n"));
    // 1) modalidade/região
    const textoMod = [titulo, descricao, pagina].filter(Boolean).join("\n");
    if (textoMod) {
      const [ok, motivo] = vagaAceita(textoMod, perfil.modalidades_aceitas, perfil.regioes_relocacao);
      if (!ok) return _logGate(platform, titulo, false, motivo, idioma);
    }
    // 2) senioridade local (título da vaga × senioridade declarada no perfil) — ESTRICTA:
    // só passa se o nível bater EXATAMENTE ("pleno remoto" ⇒ só vaga pleno). Título sem
    // nível detectável não bloqueia aqui (o match com IA decide). Nível "gestao" (gerente/
    // coordenador/diretor) ≠ qualquer nível técnico → dev não aplica em vaga de gerente.
    const nvUser = nivelDoTexto(perfil.nivel_senioridade);
    const nvVaga = nivelDoTexto(titulo);
    if (nvUser && nvVaga && ESCADA[nvUser] !== ESCADA[nvVaga]) {
      return _logGate(platform, titulo, false, `senioridade/função: candidato ${nvUser}, vaga ${nvVaga}`, idioma);
    }
    // 3) match IA (currículo × descrição) — roda MESMO sem descrição raspada: usa
    // título + texto da página. Antes, descrição vazia = SEM match nenhum → vaga de
    // gerente passava direto ("aplicava para vaga de gerente sendo desenvolvedor").
    if ((descricao || titulo) && perfil.resumo_curriculo) {
      const limiar = (cfg.plataformas?.[platform] || {}).limiarMatch || 0;
      const descIA = descricao || [titulo, pagina].filter(Boolean).join("\n").slice(0, 2500);
      const m = await OA.bg({ type: "brain.match", payload: { descricao: descIA, titulo, empresa } });
      // IA INDISPONÍVEL não pode passar em silêncio (parecia "não chama a IA"): avisa no
      // console da aba com o motivo real (chave, créditos, timeout…).
      if (m?.ok && /indispon|fail-open|erro no match/i.test(m.motivo || "")) {
        try { console.warn(`[OA-IA] match indisponível p/ "${(titulo || "").slice(0, 50)}": ${m.motivo}`); } catch (_) {}
      }
      if (m?.ok && (m.aplicar === false || (limiar > 0 && typeof m.nota === "number" && m.nota < limiar))) {
        return _logGate(platform, titulo, false, `sem match: ${m.motivo || ""} (nota ${m.nota})`, idioma);
      }
    }
    return _logGate(platform, titulo, true, "", idioma);
  }
  // Log de TODOS os vereditos do gate no console da página — facilita ver POR QUE uma
  // vaga foi aplicada ou pulada (F12 → Console → filtro "OA-gate").
  function _logGate(platform, titulo, ok, motivo, idioma) {
    try { console.log(`[OA-gate] ${platform || "?"} | ${(titulo || "(sem título)").slice(0, 60)} → ${ok ? "APLICA" : "PULA"}${motivo ? " | " + motivo : ""}`); } catch (_) {}
    return { aplicar: ok, motivo, idioma };
  }
  OA.deveAplicar = deveAplicar;

  // ── Pré-gate por TÍTULO (IA) na COLETA da lista ─────────────────────────────
  // Antes de enfileirar os cards, cada TÍTULO vai pra IA (brain.title): faz sentido
  // com o perfil (área/senioridade)? É a "consulta antes de aplicar" — sem ela as
  // plataformas abriam e preenchiam QUALQUER vaga. Cache por sessão + pool de 3
  // chamadas em paralelo (25 cards × ~2s em série = lento demais).
  const _tituloCache = new Map();
  async function tituloOk(titulo, opts = {}) {
    const key = norm(titulo);
    if (!key) return true; // sem título → fail-open
    if (_tituloCache.has(key)) return _tituloCache.get(key);
    let ok = true;
    try {
      const r = await OA.bg({ type: "brain.title", payload: { titulo: key, empresa: opts.empresa || "" } });
      ok = !r || r.aplicar !== false; // SW falhou → fail-open
    } catch (_) { /* fail-open */ }
    _tituloCache.set(key, ok);
    return ok;
  }
  async function filtrarTitulos(itens, onProgresso) {
    // itens: [{ titulo, empresa?, ref }] → devolve os aprovados (ordem preservada)
    const aprovados = [];
    if (!itens.length) return aprovados;
    let i = 0;
    const worker = async () => {
      while (i < itens.length) {
        const it = itens[i++];
        if (await tituloOk(it.titulo, it)) aprovados.push(it);
        if (onProgresso && i % 5 === 0) onProgresso(i, itens.length);
      }
    };
    await Promise.all(Array.from({ length: Math.min(3, itens.length) }, worker));
    return aprovados;
  }
  OA.tituloOk = tituloOk;
  OA.filtrarTitulos = filtrarTitulos;
})();
